benches/hashmap_benchmark.rs

```diff
@@ -0,0 +1,228 @@
+// In the search benches, we randomly search for elements of a range of SEARCH_SIZE.0/SEARCH_SIZE.1
+// times the number of elements contained.
+const SEARCH_SIZE: (usize, usize) = (4, 3);
+
+pub fn insert_empty_value(c: &mut Criterion) {
+    c.bench_function("insert_empty_value", |b| {
+        b.iter_batched(
+            || {
+                let mut rng = thread_rng();
+                let count = rng.gen_range(INSERT_COUNT.0, INSERT_COUNT.1);
+                let mut list = Vec::with_capacity(count);
+                for _ in 0..count {
+                    list.push(rng.gen_range(0, INSERT_COUNT.1 << 8) as u32)
+                }
+                (HashMap::new(), list)
+            },
+            |mut data| insert_vec(&mut data),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn insert_struct_value(c: &mut Criterion) {
+    c.bench_function("insert_struct_value", |b| {
+        b.iter_batched(
+            || {
+                let mut rng = thread_rng();
+                let count = rng.gen_range(INSERT_COUNT.0, INSERT_COUNT.1);
+                let mut list = Vec::with_capacity(count);
+                for _ in 0..count {
+                    list.push((
+                        rng.gen_range(0, INSERT_COUNT.1 << 8) as u32,
+                        default_struct(),
+                    ))
+                }
+                (HashMap::<u32, Struct>::new(), list)
+            },
+            |mut data| insert_struct_vec(&mut data),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_empty_value(c: &mut Criterion) {
+    c.bench_function("remove_empty_value", |b| {
+        b.iter_batched(
+            || prepare_remove(()),
+            |mut data| remove_vec(&mut data),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_struct_value_no_read(c: &mut Criterion) {
+    c.bench_function("remove_struct_value_no_read", |b| {
+        b.iter_batched(
+            || prepare_remove(Struct::default()),
+            |mut data| remove_vec(&mut data),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn search_empty_value(c: &mut Criterion) {
+    c.bench_function("search_empty_value", |b| {
+        b.iter_batched(
+            || prepare_search(()),
+            |data| search_vec(&data),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn search_struct_value(c: &mut Criterion) {
+    c.bench_function("search_struct_value", |b| {
+        b.iter_batched(
+            || prepare_search(Struct::default()),
+            |data| search_vec(&data),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+criterion_group!(insert, insert_empty_value, insert_struct_value);
+criterion_group!(remove, remove_empty_value, remove_struct_value_no_read);
+criterion_group!(search, search_empty_value, search_struct_value);
+criterion_main!(insert, remove, search);
+
+fn insert_vec(pair: &mut (HashMap<u32, ()>, Vec<u32>)) {
```

Comment:
These should be renamed to insert_vec_rollback due to the lack of commit. There should be a seperate type that commits after the insert.